packages/fuchsia_ctl/lib/src/emulator.dart

```diff
@@ -0,0 +1,232 @@
+
+import 'dart:io';
+
+import 'package:file/file.dart';
+import 'package:file/local.dart';
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as path;
+import 'package:process/process.dart';
+
+import 'operation_result.dart';
+
+/// A wrapper for running Fuchsia images on the Android Emulator (AEMU).
+class Emulator {
+  /// Creates a new wrapper for the `emu` tool.
+  Emulator({
+    @required this.aemuPath,
+    @required this.fuchsiaImagePath,
+    @required this.fuchsiaSdkPath,
+    this.fs = const LocalFileSystem(),
+    this.processManager = const LocalProcessManager(),
+    @required this.qemuKernelPath,
+    this.sshPath = '.fuchsia',
+    @required this.zbiPath,
+  }) : assert(processManager != null);
+
+  /// The path to the AEMU executable on disk.
+  final String aemuPath;
+
+  /// Fuchsia image to load into the emulator.
+  final String fuchsiaImagePath;
+
+  /// The path to the Fuchsia SDK that contains the tools `fvm` and `zbi`.
+  final String fuchsiaSdkPath;
+
+  /// The QEMU kernel image to use. This is only bundled in Fuchsia QEMU images.
+  final String qemuKernelPath;
+
+  /// The path to the directory containing authorized_keys.
+  final String sshPath;
+
+  /// The Fuchsia bootloader image.
+  final String zbiPath;
+
+  /// Location of `fvm` in [fuchsiaSdkPath].
+  @visibleForTesting
+  final String fvmToolPath = 'sdk/tools/fvm';
+
+  /// Location of `zbi` in [fuchsiaSdkPath].
+  @visibleForTesting
+  final String zbiToolPath = 'sdk/tools/zbi';
+
+  /// Default AEMU window size to be launched.
+  @visibleForTesting
+  final String defaultWindowSize = '1280x800';
+
+  /// Flag to pass to AEMU to run in headless mode.
+  @visibleForTesting
+  final String aemuHeadlessFlag = '-no-window';
+
+  /// The [FileSystem] to use when running the `emu` tool.
+  final FileSystem fs;
+
+  /// The [ProcessManager] to use for running the `emu` tool.
+  final ProcessManager processManager;
+
+  /// FVM extended image of [fuchsiaImagePath] for running on FEMU.
+  @visibleForTesting
+  String fvmImagePath;
+
+  /// [zbiPath] that is accessible with SSH using [sshPath] keys.
+  @visibleForTesting
+  String signedZbiPath;
+
+  /// Update given Fuchsia assets to make them compatible with FEMU.
+  ///
+  /// 1. Ensure required assets exist.
+  /// 2. Create FVM image for running with FEMU.
+  /// 3. Sign boot image for host access to the guest FEMU instance.
+  Future<void> prepareEnvironment() async {
+    assert(fs.isFileSync(fuchsiaImagePath));
+    assert(fs.isFileSync(zbiPath));
+    assert(fs.isFileSync(qemuKernelPath));
+
+    final String tmpPath = fs.systemTempDirectory.createTempSync().path;
+    fvmImagePath = '$tmpPath/fvm.blk';
+    signedZbiPath = '$tmpPath/fuchsia-ssh.zbi';
+
+    await _prepareFvmImage(fuchsiaImagePath, fvmImagePath);
+    await _signBootImage(zbiPath, signedZbiPath);
+  }
+
+  /// Double the size of [fuchsiaImagePath] to make space for the emulator
+  /// to write back to it.
```

Comment:
This is for when applications or tests are run on the emulator, the emulator has space to write back to to add those files. This could be extended to take a precise value to mock devices better, but this is the current logic `fx emu` does.